### ISSUE TITLE
Unable to run existing android app after upgrading to 1.6.0

### ISSUE THREAD

**Issue by enchev** [NativeScript/NativeScript#1577]
_From @Daxito on February 17, 2016 16:30_

I just upgraded to 1.6.0 from 1.5.2 and now I get this error when trying to run the app on android, ios is ok:
Any clue? If I create a new project it works ok, seems like a problem with my existing app only, this is my exception log:

java.lang.RuntimeException: Unable to create application com.tns.NativeScriptApplication: com.tns.NativeScriptException: 
Calling js method onCreate failed

Error: java.lang.RuntimeException: NativeScriptApplication already initialized
    com.tns.Platform.init(Platform.java:105)
    com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:1033)
    com.tns.Platform.callJSMethodNative(Native Method)
    com.tns.Platform.dispatchCallJSMethodNative(Platform.java:812)
    com.tns.Platform.callJSMethod(Platform.java:711)
    com.tns.Platform.callJSMethod(Platform.java:690)
    com.tns.Platform.callJSMethod(Platform.java:680)
    com.tns.NativeScriptApplication.onCreateInternal(NativeScriptApplication.java:924)
    com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:1037)
    android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1020)
    android.app.ActivityThread.handleBindApplication(ActivityThread.java:5010)
    android.app.ActivityThread.access$1600(ActivityThread.java:172)
    android.app.ActivityThread$H.handleMessage(ActivityThread.java:1482)
    android.os.Handler.dispatchMessage(Handler.java:102)
    android.os.Looper.loop(Looper.java:145)
    android.app.ActivityThread.main(ActivityThread.java:5835)
    java.lang.reflect.Method.invoke(Native Method)
    java.lang.reflect.Method.invoke(Method.java:372)
    com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
    com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
File: "<unknown>, line: 0, column: -1

StackTrace: 

at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5013)
at android.app.ActivityThread.access$1600(ActivityThread.java:172)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1482)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:145)
at android.app.ActivityThread.main(ActivityThread.java:5835)
at java.lang.reflect.Method.invoke(Native Method)
at java.lang.reflect.Method.invoke(Method.java:372)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
Caused by: com.tns.NativeScriptException: 
Calling js method onCreate failed

Error: java.lang.RuntimeException: NativeScriptApplication already initialized
    com.tns.Platform.init(Platform.java:105)
    com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:1033)
    com.tns.Platform.callJSMethodNative(Native Method)
    com.tns.Platform.dispatchCallJSMethodNative(Platform.java:812)
    com.tns.Platform.callJSMethod(Platform.java:711)
    com.tns.Platform.callJSMethod(Platform.java:690)
    com.tns.Platform.callJSMethod(Platform.java:680)
    com.tns.NativeScriptApplication.onCreateInternal(NativeScriptApplication.java:924)
    com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:1037)
    android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1020)
    android.app.ActivityThread.handleBindApplication(ActivityThread.java:5010)
    android.app.ActivityThread.access$1600(ActivityThread.java:172)
    android.app.ActivityThread$H.handleMessage(ActivityThread.java:1482)
    android.os.Handler.dispatchMessage(Handler.java:102)
    android.os.Looper.loop(Looper.java:145)
    android.app.ActivityThread.main(ActivityThread.java:5835)
    java.lang.reflect.Method.invoke(Native Method)
    java.lang.reflect.Method.invoke(Method.java:372)
    com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
    com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
File: "<unknown>, line: 0, column: -1

StackTrace: 

at com.tns.Platform.callJSMethodNative(Native Method)
at com.tns.Platform.dispatchCallJSMethodNative(Platform.java:812)
at com.tns.Platform.callJSMethod(Platform.java:711)
at com.tns.Platform.callJSMethod(Platform.java:690)
at com.tns.Platform.callJSMethod(Platform.java:680)
at com.tns.NativeScriptApplication.onCreateInternal(NativeScriptApplication.java:924)
at com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:1037)
at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1020)
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5010)
... 9 more
Caused by: java.lang.RuntimeException: NativeScriptApplication already initialized
at com.tns.Platform.init(Platform.java:105)
at com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:1033)
... 18 more

_Copied from original issue: NativeScript/nativescript-cli#1497_

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 17:1_

Did you upgrade BOTH the core modules, and the runtime?

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 17:5_

I followed these instructions:

https://docs.nativescript.org/releases/upgrade-instructions

npm install -g nativescript
tns platform remove android
tns platform add android
tns platform remove ios
tns platform add ios
npm install tns-core-modules@latest --save

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 17:6_

Yep; those are the correct instructions.  I'm currently in the process of upgrading one of mine.  (Slow internet right now for some reason).   I'll let you know how it goes...

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 17:13_

Interesting, I have duplicated this issue on my own project.  Investigating...

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 17:23_

I see, thanks!

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 18:20_

Are you loading / running any plugins inside the app.js?

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 18:45_

Yes, under the node_modules I have lodash, moment, nativescript-drop-down, nativescript-sqlite, nativescript-telerik-ui, numeral, string.

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 18:48_

oh wait, inside the app.js
a module that references
var connectivity = require("connectivity");
var dialogs = require("ui/dialogs"); 
var moment = require("moment");
var _ = require("lodash");
var numeral = require("numeral");

And directly under app.js:
var application = require("application");
var Everlive = require("~/lib/everlive.all.min");

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 18:51_

Can you post your app.js file.  Your comment above left me confused.  :grinning:

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 18:54_

here it is:

var application = require("application");
var converters = require('~/converters/converters');
var connectionMonitor = require('~/service/connection-monitor');
var Everlive = require("~/lib/everlive.all.min");

application.cssFile = "./app.css";
application.mainModule = "./views/login/login";

application.resources["dateConverter"] = converters.dateConverter;
application.resources["visibilityConverter"] = converters.visibilityConverter;
application.resources["idToNameConverter"] = converters.idToNameConverter;
application.resources["timeConverter"] = converters.timeConverter;
application.resources["numberConverter"] = converters.numberConverter;

global.TOKEN_DATA_KEY = "authenticationToken"; 
global.USER_ITEM = "useritem";
global.SELECT_DEFAULT_TEXT = "[Select...]"
global.CONNECTION_MONITOR = connectionMonitor;
global.CONNECTION_MONITOR.startMonitoring();


// Initialize Everlive SDK
global.EVERLIVE = new Everlive(global.BS_API_KEY);

global.COMPANY_CODE = null;

application.onSuspend = function () {
    console.log("Application suspended.");
};

application.onResume = function () {
    console.log("Application resumed.");
};

application.onExit = function () {
    console.log("Application will exit.");   
};

application.onLowMemory = function () {
    console.log("Memory is low.");
};

application.onUncaughtError = function (error) {
    console.log("Application error: " + error.name + "; " + error.message + "; " + error.nativeError);
};

application.start();

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 18:56_

Converters content is:

var moment = require("moment");
var _ = require("lodash");
var numeral = require("numeral");

var visibilityConverter = function (value) { //To View Converter    
    if (value) {
        return 'visible';
    } else {
        return 'collapsed';
    }
}

var numberConverter = function (value, format) {
    if (!_.isNull(value)) {
        format = format || "0,0.00";
        if (_.isNumber(value)) {
            return numeral(value).format(format);
        } else {
            return '';
        }
    } else {
        return '';
    }
}

var dateConverter = function (value, format) {
    if (!_.isNull(value)) {
        var m = moment(value);
        format = format || "L";
        if (m.isValid()) {
            return m.format(format);
        } else {
            return '';
        }
    } else {
        return '';
    }
}

var timeConverter = function (value, format) {
    if (!_.isNull(value) && !_.isUndefined(value)) {        
        var m = moment(value);
        format = format || "LT";
        if (m.isValid()) {             
             return m.format(format);
        } else {
            return '';
        }
    } else {
        return '';
    }
}

var idToNameConverter = function (value, items, defaultText) {
    //We assume that the items collection contains items like {id:1,name:'foo'}
    var _defaultText = defaultText || '[Select...]';
    if (!_.isNull(value)) {
        var item = _.find(items, { id: value });
        if (item) {
            return item.name;
        } else {
            return _defaultText;
        }
    } else {
        return _defaultText;
    }
}

exports.visibilityConverter = visibilityConverter;
exports.dateConverter = dateConverter;
exports.timeConverter = timeConverter;
exports.idToNameConverter = idToNameConverter;
exports.numberConverter = numberConverter;

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 18:57_

connection-monitor content is:

var connectivity = require("connectivity");
var dialogs = require("ui/dialogs");

module.exports = (function () {
    console.log('Connection Monitor Module Initialized');
    var isMonitoring = false;
    var connectionType = connectivity.getConnectionType(); //Get the initial value on initialization.
   
    //Report the connection to the console for devs
    switch (connectionType) {
        case connectivity.connectionType.none:
            console.log("No connection");
            break;
        case connectivity.connectionType.wifi:
            console.log("WiFi connection");
            break;
        case connectivity.connectionType.mobile:
            console.log("Mobile connection");
            break;
    }

    function startMonitoring() {
        console.log('Connection Monitor Enabled');
        isMonitoring = true;
        connectivity.startMonitoring(function onConnectionTypeChanged(newConnectionType) {
            if (connectionType !== newConnectionType) {
                connectionType = newConnectionType;
                var msg = '';
                switch (newConnectionType) {
                    case connectivity.connectionType.none:
                        msg = "Lost Network Connection."
                        break;
                    case connectivity.connectionType.wifi:
                        msg = "Connected to WiFi.";
                        break;
                    case connectivity.connectionType.mobile:
                        msg = "Connected to Mobile.";
                        break;
                }

                console.log(msg);
           
                console.log("New connection type value" + connectionType);
                dialogs.alert(msg);
            }

        });
    }


    function stopMonitoring() {
        connectivity.stopMonitoring();
        isMonitoring = false;
        console.log('Connection Monitor Disabled');
    }

    return {
        connectionType:function(){return connectionType},
        stopMonitoring: stopMonitoring,
        startMonitoring: startMonitoring,
        isConnected: function(){
                return  (connectivity.getConnectionType() !== connectivity.connectionType.none)
            },
        isMonitoring: isMonitoring
    }

})();

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 19:6_

Ok, I want you to try something. 

  It will break your app later I'm sure; but lets see if it actually runs.
Comment out:
var Everlive = require("~/lib/everlive.all.min");
and
global.EVERLIVE = new Everlive(global.BS_API_KEY);

If that doesn't work try commenting out:
var connectionMonitor = require('~/service/connection-monitor');
and
//  global.CONNECTION_MONITOR = connectionMonitor;
// global.CONNECTION_MONITOR.startMonitoring();

If that doesn't work, comment out both EverLive and Connection_Monitor .

**Comment by enchev** [comment]
_From @rosen-vladimirov on February 17, 2016 19:11_

Guys, can you try removing these lines 
```JavaScript
application.cssFile = "./app.css";
application.mainModule = "./views/login/login";
```
and replace the start method with this one:
```JavaScript
application.start({ moduleName: "./views/login/login" });
```

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 19:13_

Let me try all the recommendations right now.

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 19:26_

@rosen-vladimirov In my specific case switching the cssFile/MainModule to the new way didn't help.   In fact I did quite a few things to my project and finally narrowed it down to the actual context access.   I would not be surprised if something in his app is doing the same thing.

**Comment by enchev** [comment]
_From @rosen-vladimirov on February 17, 2016 19:27_

ok, I'll try to reproduce it on my side as well. Can you guide me for something specific. Why your app worked with 1.5.2 and not with 1.6.0

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 19:30_

OK, so this is the cause of the issue in my case, this is what gets called from the app.js when the module is "required", as you can see:  

var connectivity = require("connectivity");
connectivity.getConnectionType();

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 19:31_

And this is a copy of my manifest, just in case:

<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
	package="__PACKAGE__"
	android:versionCode="1"
	android:versionName="1.0">

	<supports-screens
		android:smallScreens="true"
		android:normalScreens="true"
		android:largeScreens="true"
		android:xlargeScreens="true"/>

	<uses-sdk
		android:minSdkVersion="17"
		android:targetSdkVersion="__APILEVEL__"/>

	<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
	<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
	<uses-permission android:name="android.permission.INTERNET"/>
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
    
	<application
		android:name="com.tns.NativeScriptApplication"
		android:allowBackup="true"
		android:icon="@drawable/icon"
		android:label="@string/app_name"
		android:theme="@style/AppTheme" >
		<activity
			android:name="com.tns.NativeScriptActivity"
			android:label="@string/title_activity_kimera"
			android:configChanges="keyboardHidden|orientation|screenSize">

				<intent-filter>
				<action android:name="android.intent.action.MAIN" />

				<category android:name="android.intent.category.LAUNCHER" />
			</intent-filter>
		</activity>
		<activity android:name="com.tns.ErrorReportActivity"/>
	</application>
</manifest>

**Comment by enchev** [comment]
_From @rosen-vladimirov on February 17, 2016 19:32_

This is the manifest from `app/App_Resources/Android` right?

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 19:34_

Correct, this is the new default manifest generated, I just added: 
uses-permission android:name="android.permission.ACCESS_NETWORK_STATE

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 19:36_

@rosen-vladimirov Here is the key why it is crashing so you don't have to debug it...
```
function getConnectivityManager() {
    return utils.ad.getApplicationContext().getSystemService(android.content.Context.CONNECTIVITY_SERVICE);
}
```
See the **utils.ad.getApplicationContext()**  -- the minute you get the Application context without the Application being started it crashed.  In my code Either of these two lines will cause it to crash:

```
    var ctx = java.lang.Class.forName("android.app.AppGlobals").getMethod("getInitialApplication", null).invoke(null, null);
    var ctx = java.lang.Class.forName("android.app.ActivityThread").getMethod("currentApplication", null).invoke(null, null);
```

I have a routine that initializes a plugin; if the {N} application.context isn't yet set then it uses those two methods listed above to get the actual context.  This has worked since v1.2 when I first wrote the code.

**Comment by enchev** [comment]
_From @rosen-vladimirov on February 17, 2016 19:41_

Great work @NathanaelA , 
Soooo, in case you move the `application.start()` call before your lines, the app should work?

**Comment by enchev** [comment]
_From @rosen-vladimirov on February 17, 2016 19:41_

We'll have to check this with the runtime and modules teams. @slavchev , @hshristov any thoughts on this?

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 19:47_

Yes @rosen-vladimirov , it works if I move the start before.
Thank you all! I can use this workaround until you fix it in case it is actually broken.

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 19:49_

Yes.  However, on iOS once you do an application.start() nothing below it is ran.  ;-(

Please note, I did a quick search of the tns-core-modules for getApplicationContext() and it is used in several places like filesystem, and platform.  The good news is that this bug shouldn't be too wide spread as far as I can tell.  Most people probably leave their app.js file alone...

**Comment by enchev** [comment]
_From @Daxito on February 17, 2016 19:53_

Well, I can also move that to my home page instead.

**Comment by enchev** [comment]
_From @NathanaelA on February 17, 2016 20:54_

Here is a workaround I'm using in my app:
```
var application = require("application");
if (global.android) {
    application.start({moduleName: "main-page"});
}

// ALL my custom code including any additional requires...

if (!global.android) {
    application.start({moduleName: "main-page"});
}
```

**Comment by enchev** [comment]
Hey folks,

I've just submitted pull request related to this issue. The error in this case was caused by the fact that `connectivity.getConnectionType()` will try access Android application instance however the application is not yet started - this can be avoided if you call the method using `launch` event. 

Let me know what you think!

Thanks

**Comment by NathanaelA** [comment]
@enchev I have a couple issues with your proposed solution.


1. These two different method may be present in _native_ plugins that may initialize at startup that need the app context...   
```
  var ctx = java.lang.Class.forName("android.app.AppGlobals").getMethod("getInitialApplication", null).invoke(null, null);
```
```
    var ctx = java.lang.Class.forName("android.app.ActivityThread").getMethod("currentApplication", null).invoke(null, null);
```
Both of these lines of code if ran before application.start() will cause the exact same crashing error.


\2.   What happens if I want to do something like:
```
var application = require("application");
var platform = require('platform');

var startupInfo = {moduleName: "main-page"};

if (platform.device.deviceType === "Tablet") {
    startupInfo.moduleName = "main-page-tablet";
}
application.start(startupInfo);
```

or

```
var application = require("application");
var appSettings = require('application-settings');

var startupInfo = {moduleName: "main-page-normal"};

if (appSettings.getBoolean('useAdvancedInterface')) {
   startupInfo.moduleName = "main-page-advanced";
}
application.start(startupInfo);
```

ALL of these scenarios work fine in 1.0.0 through 1.5.x

**Comment by enchev** [comment]
Hey @slavchev, @atanasovg Can you help here?

**Comment by atanasovg** [comment]
@NathanaelA You are correct - this used to work prior 1.6. The behavioral change is caused by the [snapshot refactoring pull](https://github.com/NativeScript/NativeScript/pull/1407) because the `app.init` object is now populated in the `application.start` method - see https://github.com/NativeScript/NativeScript/blob/master/application/application.android.ts#L300-L311. In previous versions of the modules, if you access the Android Application object before requiring the `application` module you will get the very same exception. Why we did this change - because V8 may not serialize external memory and everything the Runtime injects as proxies and callbacks needs to happen after the snapshot.

This exception is generally a quirk in the Runtime itself and we may have this behavior improved by simply tracking the initialization process via some `initializing` flag. Good news is that for 1.7 we will do our best to integrate the new Activity/Application resolution (https://github.com/NativeScript/android-runtime/issues/283) and the problem will not exist with the new functionality.

Now, I guess I haven't estimated well the impact of this change - my thinking was that it shouldn't be hit by the users and obviously my assumption was wrong. Your scenarios are perfectly valid and if this turns to be a show-stopper we will definitely push a hot fix for it.

**Comment by NathanaelA** [comment]
@atanasovg - fortunately there are some work around for most of it.   So I think just leave this bug open and see if their is anyone else who screams bloody murder...  If it is fixed in v1.7 -- I'm personally good to wait for this one.

**Comment by atanasovg** [comment]
https://github.com/NativeScript/android-runtime/commit/a89aa468ec27b8f0660531520fe97c635b02ea18 fixes the issue, we need to prepare a hot fix for the Android Runtime.
